File: backends/python/hypergrid/casbin_logger.py
"""
HyperGrid Casbin Logger — optional authorization logging.

Usage:
    logger = CasbinLogger(enabled=True, min_level="info")
    grid_engine = HyperGridJinjaEngine(enforcer, logger=logger)
"""

import json
import logging
import sys
from datetime import datetime
from pathlib import Path
from typing import Optional


class CasbinLogger:
    """Optional logger for Casbin authorization decisions.

    Supports multiple output handlers: file, stdout, stderr.
    Log levels: debug, info, notice, warning, error, critical.
    """

    LEVELS = {
        "debug": 0,
        "info": 1,
        "notice": 2,
        "warning": 3,
        "error": 4,
        "critical": 5,
    }

    def __init__(
        self,
        enabled: bool = False,
        min_level: str = "info",
        log_file: Optional[str] = None,
        handlers: Optional[list] = None,
    ):
        self.enabled = enabled
        self.min_level = self.LEVELS.get(min_level, 1)
        self.log_file = log_file or str(
            Path(Path.home(), ".hypergrid", "casbin.log")
        )
        self.handlers = handlers or ["file"]

        if self.enabled:
            log_dir = Path(self.log_file).parent
            log_dir.mkdir(parents=True, exist_ok=True)

        # Python stdlib logger as secondary output
        self._py_logger = logging.getLogger("hypergrid.casbin")
        self._py_logger.setLevel(logging.DEBUG if enabled else logging.WARNING)

    def _should_log(self, level: str) -> bool:
        return self.enabled and self.LEVELS.get(level, 0) >= self.min_level

    def _write(self, level: str, message: str, context: Optional[dict] = None):
        if not self._should_log(level):
            return

        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
        ctx_str = ""
        if context:
            ctx_str = " " + json.dumps(context, default=str)
        entry = f"[{timestamp}] [hypergrid.casbin.{level}] {message}{ctx_str}\n"

        for handler in self.handlers:
            if handler == "file":
                try:
                    with open(self.log_file, "a", encoding="utf-8") as f:
                        f.write(entry)
                except OSError:
                    pass
            elif handler == "stdout":
                sys.stdout.write(entry)
                sys.stdout.flush()
            elif handler == "stderr":
                sys.stderr.write(entry)
                sys.stderr.flush()

        # Also pipe to stdlib logger
        log_method = getattr(self._py_logger, level, self._py_logger.info)
        log_method("%s %s", message, context or {})

    def debug(self, message: str, context: Optional[dict] = None):
        self._write("debug", message, context)

    def info(self, message: str, context: Optional[dict] = None):
        self._write("info", message, context)

    def warning(self, message: str, context: Optional[dict] = None):
        self._write("warning", message, context)

    def error(self, message: str, context: Optional[dict] = None):
        self._write("error", message, context)

    def critical(self, message: str, context: Optional[dict] = None):
        self._write("critical", message, context)

    def log_enforce(self, user_id: str, obj: str, action: str, allowed: bool):
        """Log an authorization enforcement decision."""
        level = "info" if allowed else "warning"
        self._write(
            level,
            "Enforce decision",
            {"user": user_id, "object": obj, "action": action, "allowed": allowed},
        )

    def log_cell_access(self, user_id: str, col_name: str, action: str, granted: bool):
        """Log a cell-level access check."""
        level = "debug" if granted else "warning"
        self._write(
            level,
            "Cell access check",
            {"user": user_id, "column": col_name, "action": action, "granted": granted},
        )

File: backends/go/src/casbin_logger.go
package hypersheet

import (
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"
)

// LogLevel represents the severity of a log entry
type LogLevel int

const (
	DebugLevel LogLevel = iota
	InfoLevel
	NoticeLevel
	WarningLevel
	ErrorLevel
	CriticalLevel
)

func (l LogLevel) String() string {
	switch l {
	case DebugLevel:
		return "debug"
	case InfoLevel:
		return "info"
	case NoticeLevel:
		return "notice"
	case WarningLevel:
		return "warning"
	case ErrorLevel:
		return "error"
	case CriticalLevel:
		return "critical"
	default:
		return "unknown"
	}
}

// CasbinLogger provides optional logging for Casbin authorization decisions
type CasbinLogger struct {
	mu       sync.Mutex
	enabled  bool
	level    LogLevel
	logFile  string
	handlers []string
	logger   *log.Logger
}

// NewCasbinLogger creates a new Casbin logger. Set enabled=false to disable.
func NewCasbinLogger(config map[string]string) *CasbinLogger {
	cl := &CasbinLogger{
		enabled:  config["enabled"] == "true",
		level:    parseLevel(config["min_level"]),
		logFile:  config["log_file"],
		handlers: strings.Split(config["handlers"], ","),
	}

	if cl.logFile == "" {
		cl.logFile = "/tmp/Hypersheet_casbin.log"
	}
	if len(cl.handlers) == 0 || cl.handlers[0] == "" {
		cl.handlers = []string{"file"}
	}

	if cl.enabled {
		f, err := os.OpenFile(cl.logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err == nil {
			cl.logger = log.New(f, "", 0)
		}
	}

	return cl
}

func parseLevel(s string) LogLevel {
	switch strings.ToLower(s) {
	case "debug":
		return DebugLevel
	case "info":
		return InfoLevel
	case "notice":
		return NoticeLevel
	case "warning":
		return WarningLevel
	case "error":
		return ErrorLevel
	case "critical":
		return CriticalLevel
	default:
		return InfoLevel
	}
}

func (cl *CasbinLogger) shouldLog(level LogLevel) bool {
	return cl.enabled && level >= cl.level
}

func (cl *CasbinLogger) write(level LogLevel, msg string, ctx map[string]interface{}) {
	if !cl.shouldLog(level) {
		return
	}

	cl.mu.Lock()
	defer cl.mu.Unlock()

	timestamp := time.Now().Format("2006-01-02 15:04:05.000")
	ctxStr := ""
	if len(ctx) > 0 {
		ctxStr = " " + fmt.Sprintf("%v", ctx)
	}
	entry := fmt.Sprintf("[%s] [Hypersheet.casbin.%s] %s%s\n", timestamp, level, msg, ctxStr)

	for _, h := range cl.handlers {
		switch strings.TrimSpace(h) {
		case "file":
			if cl.logger != nil {
				cl.logger.Print(entry)
			}
		case "stdout":
			fmt.Print(entry)
		case "stderr":
			fmt.Fprint(os.Stderr, entry)
		}
	}
}

// Debug logs at debug level
func (cl *CasbinLogger) Debug(msg string, ctx map[string]interface{}) {
	cl.write(DebugLevel, msg, ctx)
}

// Info logs at info level
func (cl *CasbinLogger) Info(msg string, ctx map[string]interface{}) {
	cl.write(InfoLevel, msg, ctx)
}

// Warning logs at warning level
func (cl *CasbinLogger) Warning(msg string, ctx map[string]interface{}) {
	cl.write(WarningLevel, msg, ctx)
}

// Error logs at error level
func (cl *CasbinLogger) Error(msg string, ctx map[string]interface{}) {
	cl.write(ErrorLevel, msg, ctx)
}

// LogEnforce logs an authorization decision
func (cl *CasbinLogger) LogEnforce(userID, object, action string, allowed bool) {
	level := InfoLevel
	if !allowed {
		level = WarningLevel
	}
	cl.write(level, "Enforce decision", map[string]interface{}{
		"user":    userID,
		"object":  object,
		"action":  action,
		"allowed": allowed,
	})
}

// LogCellAccess logs a cell-level access check
func (cl *CasbinLogger) LogCellAccess(userID, colName, action string, granted bool) {
	level := DebugLevel
	if !granted {
		level = WarningLevel
	}
	cl.write(level, "Cell access check", map[string]interface{}{
		"user":    userID,
		"column":  colName,
		"action":  action,
		"granted": granted,
	})
}

// Close flushes and closes the log file
func (cl *CasbinLogger) Close() {
	cl.enabled = false
}
